.teamcity/src/main/kotlin/model/FunctionalTestBucketProvider.kt

```diff
@@ -86,10 +86,33 @@ class StatisticBasedFunctionalTestBucketProvider(val model: CIBuildModel, testBu
             val buckets: List<BuildTypeBucket> = testCoverageAndBucket.getJSONArray("buckets").map {
                 fromJsonObject(it as JSONObject).toBuildTypeBucket(model.subprojects)
             }
-            testCoverage to buckets
+
+            // Sometimes people may add new subproject into `subprojects.json`
+            // in this case we have no historical test running time, so we simply add these subprojects into first available bucket
+            val allSubprojectsInBucketJson = buckets.flatMap {
+                if (it is SmallSubprojectBucket) it.subprojects.map { it.name }
+                else listOf((it as LargeSubprojectSplitBucket).subproject.name)
+            }.toSet()
+            val allSubprojectsInModel = model.subprojects.subprojects.filter { it.functionalTests || it.unitTests || it.crossVersionTests }.map { it.name }.toMutableList()
+            allSubprojectsInModel.removeAll(allSubprojectsInBucketJson)
+
+            if (buckets.isEmpty() || allSubprojectsInModel.isEmpty()) {
```

Comment:
How can it be possible that `buckets` is empty? Why are we even here if we don't want to generate tests for this stage?

Reply:
I'm keeping the behavior same as before, because there're several empty buckets for soak tests:

```
test-buckets.json:

	{
		"buckets":[],
		"testCoverageUuid":8
	},
	{
		"buckets":[],
		"testCoverageUuid":9
	},
	{
		"buckets":[],
		"testCoverageUuid":35
	},
```

```
CIBuildModel.kt:

                TestCoverage(8, TestType.soak, Os.LINUX, JvmCategory.MAX_VERSION, 1),
                TestCoverage(9, TestType.soak, Os.WINDOWS, JvmCategory.MIN_VERSION, 1),
                TestCoverage(35, TestType.soak, Os.MACOS, JvmCategory.MIN_VERSION, 1),
```

Reply:
Now I realized it shouldn't exist at all, fixed.